internal/server/deviceserver.go

```diff
@@ -32,6 +34,9 @@ func (s *deviceServer) Send(ctx context.Context, d *pb.Data) (*pb.Receipt, error
 		if err != nil {
 			log.Warnf("cannot unmarshal message. DeviceID: %s; err: %v", s.configManager.GetDeviceID(), err)
 		}
+		// Don't allow for parallel configuration updates if one of them takes longer.
+		s.updateLock.Lock()
```

Comment:
this lock will be "hard" because Update() can take a bit of time, and it's not a good way to manage it. Maybe what we need here is a waitinggroup, no?

https://gobyexample.com/waitgroups

Reply:
The whole point of that lock is to allow for executing `Update` only once at a time - allowing simultaneous execution of `Update` is a problem. We can either use a lock and allow other updates to queue on it, queue the updates ourselves (here be dragons) or discard updates that come when an earlier update is in progress.

I don't see how wait group would help here. Potentially Try Lock behavior (for example https://github.com/subchen/go-trylock) would help in releasing goroutines early but I am _very_ hesitant about using 3rd party concurrency tools or writing them myself.:)

Reply:
I see an option with atomic functions: use atomic-guarded `int32` as a try-lock:
```go
type deviceServer struct {
	locker              int32
}

func (s *deviceServer) Send(ctx context.Context, d *pb.Data) (*pb.Receipt, error) {
  ...
  if swapped := atomic.CompareAndSwapInt32(&s.locker, 0, 1); swapped {
    err = s.configManager.Update(deviceConfigurationMessage)
    atomic.StoreInt32(&s.locker, 0)
  }
  // Ignore update
  ...
  return &pb.Receipt{}, nil
}

```

Reply:
IMO, if we received a message to the device, we should process it.
With our current API, this is the behavior that we should aim for.
Unless we'll break the API into more specific actions.
The reason not to miss any update is the risk of missing an action, especially if we're going to to use device-configuration for running ansible playbooks.

Reply:
With current ansible support design we can distinguish between ansible playbook messages and configuration updates and only ignore said configuration updates. 
I would prefer to lock and process all of the updates a bit later. That said I also don't think that skipping a configuration update would be a problem - we always receive full expected configuration and eventually we will process the most recent one. 
Of course there are corner cases - someone enables some setting for a moment and then changes it back to the original setting and the temporary configuration change might be missed by the device. Taking into account that we check for configuration updates in intervals, we can face the same situation now.

IMO we can safely go either way and I lean towards my original implementation - full lock.